staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -158,3 +161,184 @@ func ValidateCustomResourceDefinitionNames(names *apiextensions.CustomResourceDe
 
 	return allErrs
 }
+
+func ValidateCustomResource(customresource interface{}, crd apiextensions.CustomResourceDefinition) error {
+	schema := spec.Schema{}
+	if err := convertToOpenAPITypes(&crd, &schema); err != nil {
+		return err
+	}
+
+	if err := spec.ExpandSchema(&schema, nil, nil); err != nil {
+		return err
+	}
+
+	validator := validate.NewSchemaValidator(&schema, nil, "spec", strfmt.Default)
+	result := validator.Validate(customresource)
+	if result.AsError() != nil {
+		return result.AsError()
+	}
+	return nil
+}
+
+func convertToOpenAPITypes(in *apiextensions.CustomResourceDefinition, out *spec.Schema) error {
+	if in.Spec.Validation != nil {
```

Comment:
@sttts re: pointer type for CustomResourceValidation

since it is optional and would need a check here to see if it exists.

On further thought, maybe a check for `in.Spec.Validation.JSONSchema` should suffice.